docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1056 @@
+
+    }
+    else if (Vector128.IsHardwareAccelerated)
+    {
+        config = config.AddJob(enough.WithId("Vector128"));
+    }
+
+    BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly)
+        .Run(args, config);
+}
+```
+
+**Note:** the config defines a [disassembler](https://adamsitnik.com/Disassembly-Diagnoser/), which exports a disassembly in GitHub markdown format (supported on both x64 and arm64, Windows and Linux). It is very often an invaluable tool when working with high-performance code where inspecting generated assembly code is required.
+
+#### Memory alignment
+
+BenchmarkDotNet does a lot of heavy lifting for the end users, but it cannot protect us from the random memory alignment which can be different per each benchmark run and can affect the stability of the benchmarks.
+
+We have three possibilities:
+
+* We can enforce the alignment ourselves and have very stable results.
+* We can ask the harness to try to randomize the memory and observe the entire possible distribution with each run.
+* We can do nothing and wonder why the results vary from time to time.
+
+##### Enforcing memory alignment
+
+We can allocate aligned unmanaged memory by using the [NativeMemory.AlignedAlloc](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.nativememory.alignedalloc).
+
+```cs
+public unsafe class Benchmarks
+{
+    private void* _pointer;
+
+    [Params(6, 32, 1024)] // test various sizes
+    public uint Size;
+
+    [GlobalSetup]
+    public void Setup()
+    {
+        _pointer = NativeMemory.AlignedAlloc(byteCount: Size * sizeof(int), alignment: 32);
+        NativeMemory.Clear(_pointer, byteCount: Size * sizeof(int)); // ensure it's all zeros, so 1 is never found
+    }
+
+    [Benchmark]
+    public bool Contains()
+    {
+        ReadOnlySpan<int> buffer = new (_pointer, (int)Size);
+        return buffer.Contains(1);
+    }
+
+    [GlobalCleanup]
+    public void Cleanup() => NativeMemory.AlignedFree(_pointer);
+}
+```
+
+Sample results (please mind the AVX2, AVX and SSE4.2 information printed in the summary):
+
+```ini
+BenchmarkDotNet=v0.13.5, OS=Windows 11 (10.0.22621.1413/22H2/2022Update/SunValley2)
+AMD Ryzen Threadripper PRO 3945WX 12-Cores, 1 CPU, 24 logical and 12 physical cores
+.NET SDK=8.0.100-alpha.1.22558.1
+  [Host]    : .NET 7.0.4 (7.0.423.11508), X64 RyuJIT AVX2
+  Scalar    : .NET 7.0.4 (7.0.423.11508), X64 RyuJIT
+  Vector128 : .NET 7.0.4 (7.0.423.11508), X64 RyuJIT AVX
+  Vector256 : .NET 7.0.4 (7.0.423.11508), X64 RyuJIT AVX2
+```
+
+```
+|   Method |       Job | Size |       Mean |    StdDev | Ratio | Code Size |
+|--------- |---------- |----- |-----------:|----------:|------:|----------:|
+| Contains |    Scalar | 1024 | 143.844 ns | 0.6234 ns |  1.00 |     206 B |
+| Contains | Vector128 | 1024 | 104.544 ns | 1.2792 ns |  0.73 |     335 B |
+| Contains | Vector256 | 1024 |  55.769 ns | 0.6720 ns |  0.39 |     391 B |
```

Comment:
For an initial example, I think it'd be worth using an example that showcases best possible results, e.g. for Int32, we'd "expect" 4x difference between Scalar and Vector256, which we're not getting here.  Otherwise, it breeds questions / confusion in the mind of a reader.  What's the reason it's only a 30% improvement rather than a 4x improvement?  Can we fix that for the example, even if it's just making the input longer?  (Especially since below you say that this is "the best case scenario".)

Reply:
> Can we fix that for the example, even if it's just making the input longer?

Sure, I'll try to come up with a better example.

Reply:
Doing `Add` between two arrays/spans is a good example. It doesn't require any fancy operations and typically hits the actual `Count * faster`

Reply:
I've been thinking about it a little bit more and I got to the following conclusion: it's not easy to find a built-in .NET method that provides a perfect perf boost thanks to vectorization. Even simple methods like `IndexOf` and `Contains` don't provide it for buffers of small primitive types.

That is why such boost might be rare and we should simply point the users to the right profilers which can tell them where the CPU time is spent.

I've added following sentece below the table with the results:

> **Note:** as you can see, even such simple method like [Contains](https://learn.microsoft.com/dotnet/api/system.memoryextensions.contains) **did not observe a perfect performance boost**: x8 for `Vector256` (256/32) and x4 for `Vector128` (128/32). To understand why, we would need to use a profiler that provides information on CPU instruction level, which depending on the hardware could be [Intel VTune](https://www.intel.com/content/www/us/en/developer/tools/oneapi/vtune-profiler.html) or [amd uprof](https://developer.amd.com/amd-uprof/).

@stephentoub @tannergooding thoughts?